### PR TITLE
[Lottie exporter] Add support for linear gradient

### PR DESCRIPTION
**Pull request by saurabh-ramola**
Linear gradients whose colors opacity is set to 100% are now working. I have raised my concerns at lottie-web repo regarding the issue I am facing(https://github.com/airbnb/lottie-web/issues/2055#issuecomment-603730904).
I have made this as a draft because the opacity of gradients is not yet supported.
@AnishGG please have a look at this. ^__^

### CODE REVIEW

File: synfig-studio/plugins/lottie-exporter/common/Gradient.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        Args:
+            (None)
+
+        Returns:
+            (list) : Stores the gradient in the list format
+        """
+        ret = []
+        for col in self.colors:
+            pos = col[0]
+            r, g, b = col[1].red, col[1].blue, col[1].green
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/common/Gradient.py]: I am getting some different results in the colors. It should have been like this:
```
r, g, b = col[1].red, col[1].green, col[1].blue
```

File: synfig-studio/plugins/lottie-exporter/shapes/rectangle.py [hunk 2]
```diff
@@ -46,6 +46,46 @@ def get_child_value(is_animate, child, what_type):
...
+    lottie["d"] = settings.DEFAULT_DIRECTION
+    lottie["s"] = {}        # Size of rectangle
+    lottie["ix"] = idx      # setting the index
+    lottie["r"] = {}        # Rounded corners of rectangle
+
+    bevel = 0
+    gen_properties_value(lottie["r"], bevel, index.inc(), settings.DEFAULT_ANIMATED, settings.NO_INFO)
+
+    # Set the position of the rectangle to the center of the Synfig canvas
+    p1, p2 = settings.lottie_format["w"] / 2, settings.lottie_format["h"] / 2
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/shapes/rectangle.py]: When the gradient is inside a group layer, you will have to take care of the size and position more effectively. `settings.ADDITIONAL_PRECOMP_WIDTH` AND `settings.ADDITIONAL_PRECOMP_HEIGHT` will now come into picture.
Have a look at these links:
https://github.com/synfig/synfig/blob/f806eb5649408a395b383e7991da6149ded94832/synfig-studio/plugins/lottie-exporter/synfig/group.py#L15
https://github.com/synfig/synfig/blob/f806eb5649408a395b383e7991da6149ded94832/synfig-studio/plugins/lottie-exporter/common/Layer.py#L147

File: synfig-studio/plugins/lottie-exporter/common/Gradient.py (hunk 1 above)
Comment (saurabh-ramola) [reply synfig-studio/plugins/lottie-exporter/common/Gradient.py]: Fixed it

**Comment by saurabh-ramola** [comment]
@AnishGG , Linear Gradient layer is now complete. Can you please test it and let me know if it is correct?

**Comment by AnishGG** [comment]
I will test this soon!

File: synfig-studio/plugins/lottie-exporter/common/Gradient.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        Returns:
+            (common.Color.Color) : color element with gamma effect reversed
+        """
+        ret = copy.deepcopy(color)
+        ret.red = ret.red ** settings.GAMMA[0]
+        ret.green = ret.green ** settings.GAMMA[1]
+        ret.blue = ret.blue ** settings.GAMMA[2]
+        return ret
+
+    def get_lottie_final_list(self):
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/common/Gradient.py:62]: I don't think this function is used anywhere, but you can keep it for future use.

File: synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py [hunk 4]
```diff
@@ -225,7 +225,7 @@ def calc_tangent(animated, lottie, i):
 
 
     ### Special case for color interpolations ###
-    if animated.attrib["type"] == "color":
+    if animated.attrib["type"] in {"color", "gradient"}:
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py:228]: Here also.

File: synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py [hunk 5]
```diff
@@ -187,7 +187,7 @@ def calc_tangent(animated, lottie, i):
 
     # After effects only supports linear,ease-in,ease-out and constant interpolations for color
     ##### No support for TCB and clamped interpolations in color is there yet #####
-    if animated.attrib["type"] == "color":
+    if animated.attrib["type"] in {"color", "gradient"}:
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py:190]: This should be linear_gradient, right?

File: synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py [hunk 6]
```diff
@@ -101,7 +101,7 @@ def gen_value_Keyframe(curve_list, animated, i):
 
     # After effects only supports linear,ease-in,ease-out and constant interpolations for color
     ##### No support for TCB and clamped interpolations in color is there yet #####
-    if animated.attrib["type"] == "color":
+    if animated.attrib["type"] == {"color", "gradient"}:
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py:104]: linear_gradient

File: synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py [hunk 7]
```diff
@@ -186,5 +186,5 @@ def set_tangents(out_val, in_val, cur_pos, next_pos, lottie, animated):
     lottie["synfig_o"] = [lottie["o"]["y"][0]]
 
     # If type is color, the tangents are already normalized
-    if animated.attrib["type"] != "color":
+    if animated.attrib["type"] not in {"color", "gradient"}:
```
Comment (AnishGG) [inline synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py:189]: linear_gradient

Comment (saurabh-ramola) [reply synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py]: fixed ^_^

File: synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py (hunk 6 above)
Comment (saurabh-ramola) [reply synfig-studio/plugins/lottie-exporter/properties/valueKeyframe.py]: fixed ^_^

File: synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py (hunk 5 above)
Comment (saurabh-ramola) [reply synfig-studio/plugins/lottie-exporter/properties/offsetKeyframe.py]: fixed ^_^

**Comment by saurabh-ramola** [comment]
@AnishGG , I have made the required changes ^_^.

**Comment by ice0** [comment]
Merged. Thank you!

**Comment by AnishGG** [comment]
Linking this to original issue: #934

**Pull request merged**